marlowe/src/Language/Marlowe/Client/History.hs

```diff
@@ -0,0 +1,335 @@
+      [creation] -> Just creation
+      _          -> Nothing
+
+
+-- | Find the Marlowe outputs at an address.
+filterOutputs :: Address            -- ^ The Marlowe address.
+              -> ChainIndexTx       -- ^ The transaction.
+              -> [MarloweTxOutRef]  -- ^ The outputs to the Marlowe address.
+filterOutputs address =
+  filter ((address ==) . txOutAddress . tyTxOutTxOut . tyTxOutRefOut)
+    . txMarloweData
+
+
+-- | Extract Marlowe data from a transaction.
+txMarloweData :: ChainIndexTx       -- ^ The transaction.
+              -> [MarloweTxOutRef]  -- ^ The outputs that have Marlowe data.
+txMarloweData citx =
+  catMaybes
+    $ uncurry (toMarlowe citx)
+    <$> txDatums citx
+
+
+-- | Make an output specific to Marlowe.
+toMarlowe :: ChainIndexTx           -- ^ The transaction.
+          -> TxOut                  -- ^ The output.
+          -> TxOutRef               -- ^ The output reference.
+          -> Maybe MarloweTxOutRef  -- ^ The Marlowe-specific output, if any.
+toMarlowe citx txOut txOutRef =
+  do
+    dh <- txOutDatumHash txOut
+    dat <- M.lookup dh $ citx ^. citxData
+    md <- fromBuiltinData $ getDatum dat
+    pure
+      . TypedScriptTxOutRef txOutRef
+      $ TypedScriptTxOut txOut md
+
+
+-- | Extract transaction outputs with datum.
+txDatums :: ChainIndexTx         -- ^ The transaction.
+         -> [(TxOut, TxOutRef)]  -- ^ The outputs that have datum.
+txDatums citx =
+  case citx ^. citxOutputs of
+    ValidTx txOuts -> filter (isJust . txOutDatumHash . fst)
+                      $ (fst &&& (TxOutRef (citx ^. citxTxId) . snd))
+                      <$> zip txOuts [0..]
+    InvalidTx      -> []
+
+
+-- | Extract Marlowe input from a transaction.
+txInputs :: ChainIndexTx                    -- ^ The transaction.
+         -> [(TxOutRef, TransactionInput)]  -- ^ The inputs that have Marlowe inputs.
+txInputs citx =
+  case citx ^. citxValidRange of
+    Interval (LowerBound (Finite l) True) (UpperBound (Finite h) False) ->
+      let
+        slots = (l, h)
+      in
+        catMaybes
+          $ secondM (fmap (TransactionInput slots) . fromBuiltinData . getRedeemer)
+          <$> txRedeemers citx
+    _ -> []  -- TODO: Should this instead throw an error in an error monad?
+
+
+-- | Extract transaction inputs with redeemers.
+txRedeemers :: ChainIndexTx            -- ^ The transaction.
+            -> [(TxOutRef, Redeemer)]  -- ^ The inputs that have redeemers.
+txRedeemers citx =
+  case citx ^. citxCardanoTx of
+    Just (SomeTx (ShelleyTx ShelleyBasedEraAlonzo tx@Alonzo.ValidatedTx{}) _) ->
+      catMaybes
+        [
+          do
+            (Alonzo.Data dat, _) <- Alonzo.indexedRdmrs tx $ Alonzo.Spending txin
+            pure (TxOutRef{..}, Redeemer . dataToBuiltinData $ dat)
+        |
+          txin@(Cardano.TxIn txid txix) <- S.toList . Alonzo.inputs . Alonzo.body $ tx
+        , let txOutRefId = TxId . toBuiltin . BS.drop 2 . toStrictByteString . toCBOR $ txid
```

Comment:
Why `drop 2`? "0x"?

Reply:
Actually, I think it is a CBOR prefix. This is how other functions in higher-level Cardano packages do the conversion, but I couldn't find an existing conversion function from the Alonzo ledger to Plutus. This was bothering me, too, so I'll dig deeper.

Reply:
Recorded as https://input-output.atlassian.net/browse/SCP-3503.